packages/backend/src/core/StateUpdater.ts

```diff
@@ -1,110 +1,54 @@
+import { ForcedAction, OraclePrice } from '@explorer/encoding'
 import {
-  ForcedAction,
-  OnChainData,
-  StarkExProgramOutput,
-} from '@explorer/encoding'
-import { PositionLeaf, RollupState } from '@explorer/state'
+  IMerkleStorage,
+  MerkleTree,
+  PositionLeaf,
+  VaultLeaf,
+} from '@explorer/state'
 import { Hash256, PedersenHash, Timestamp } from '@explorer/types'
 
 import { ForcedTransactionsRepository } from '../peripherals/database/ForcedTransactionsRepository'
-import { PageRepository } from '../peripherals/database/PageRepository'
 import { PositionRecord } from '../peripherals/database/PositionRepository'
-import { RollupStateRepository } from '../peripherals/database/RollupStateRepository'
 import { StateTransitionRecord } from '../peripherals/database/StateTransitionRepository'
 import { StateUpdateRepository } from '../peripherals/database/StateUpdateRepository'
 import { EthereumClient } from '../peripherals/ethereum/EthereumClient'
 import { BlockNumber } from '../peripherals/ethereum/types'
 import { Logger } from '../tools/Logger'
 
-/**
- * @internal
- * Same as `await RollupState.empty().then(empty => empty.positions.hash())`
- */
-export const ROLLUP_STATE_EMPTY_HASH = PedersenHash(
-  '52ddcbdd431a044cf838a71d194248640210b316d7b1a568997ecad9dec9626'
-)
-
-export class StateUpdater {
+export class StateUpdater<T extends PositionLeaf | VaultLeaf> {
   constructor(
-    private readonly pageRepository: PageRepository,
-    private readonly stateUpdateRepository: StateUpdateRepository,
-    private readonly rollupStateRepository: RollupStateRepository,
-    private readonly ethereumClient: EthereumClient,
-    private readonly forcedTransactionsRepository: ForcedTransactionsRepository,
-    private readonly logger: Logger,
-    private rollupState?: RollupState
+    protected readonly stateUpdateRepository: StateUpdateRepository,
+    protected readonly rollupStateRepository: IMerkleStorage<T>,
+    protected readonly ethereumClient: EthereumClient,
+    protected readonly forcedTransactionsRepository: ForcedTransactionsRepository,
```

Comment:
Why is ForcedTransactionsRepository plural? Every other thing is singular like StateUpdateRepository. We should probably drop the `s`. Can be in a later PR.